classlib/src/main/java/org/teavm/classlib/java/util/CollectionsFactory.java

```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright 2014 Alexey Andreev.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.teavm.classlib.java.util;
+
+import static org.teavm.classlib.java.util.TObjects.requireNonNull;
+import org.teavm.classlib.java.lang.TIllegalArgumentException;
+import org.teavm.classlib.java.lang.TNullPointerException;
+
+/**
+ * Factory-methods for List/Set/Map.of(...).
+ */
+class CollectionsFactory {
+
+    private CollectionsFactory() {
+    }
+
+    /**
+     * Create an immutable list for the {@code List.of(...)} factory methods.
+     *
+     * @throws TNullPointerException if any element is null
+     */
+    @SafeVarargs
+    static <E> TList<E> createList(E... elements) {
+        if (elements == null || elements.length == 0) {
+            return TCollections.emptyList();
+        }
+
+        final TList<E> list = new TArrayList<>();
```

Comment:
I don't like this approach:

1. Two objects created instead of one (i.e. ArrayList and UnmodifiableList)
2. You always create a temporary array for passing varargs, even if original List.of does not use varargs
3. You rely on ArrayList which can allocate extra space in order to improve `add` performance, which is not required here in immutable list.

Reply:
I wasn't sure if I should introduce new classes for these factory methods.

Do you think it's better to also have specialized implementations for the Map and Set?
THashMap and by extension THashSet aren't exactly light-weight either...

Reply:
I think that using HashMap and HashSet is ok, since their implementation is not so trivial.

Reply:
I've added a special implementation for the immutable list, which reuses the varargs-array, if possible.
Also, I think an immutable Map/Set replacement would be a lot simpler than the HashSet implementation and could be made a bit more lightweight. If you're OK with that I'd like to work on that too, but in a new PR.